### PR TITLE
✨ InaboxResources: Observe intersections for some elements' viewportCallbacks

### PR DESCRIPTION
**Pull request by alanorozco**
`viewportCallback` is not triggered from `InaboxResources`. For browsers that support `IntersectionObserver`, use that instead to trigger them.

### CODE REVIEW

**Comment by dvoytenko** [comment]
My one question: does carousel need a `viewportCallback`? Or is it only for its children?

Yeah, if we did something like this, we should look into how to do this higher up for other elements as well. But I do think the elements have to opt-in.

**Comment by alanorozco** [comment]
@dvoytenko It's to start/pause autoplay. The actual signal here is visibility, so I'm unsure what a better model would be.

**Comment by dvoytenko** [comment]
> @dvoytenko It's to start/pause autoplay. The actual signal here is visibility, so I'm unsure what a better model would be.

Got it. We need do something like this. But it looks like for now it's only needed in inabox? Can we use some API separation for inabox to limit this for now?

File: src/utils/viewport.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  }
+  let observer = new IntersectionObserver(
+    entries => {
+      entries.forEach(entry => {
+        if (entry.target === baseElement.element) {
+          baseElement.viewportCallback(entry.isIntersecting);
+        }
+      });
+    },
+    {threshold: 0.5, ...options}
```
Comment (alanorozco) [inline src/utils/viewport.js]: This threshold is the same for visibility triggers on video autoplay, so I used it for symmetry. Don't have a strong opinion on this.

**Comment by alanorozco** [comment]
@dvoytenko RE: siloing to inabox, should we try to get rid of `viewportCallback` in general in favor of `IntersectionObserver` everywhere?

This would also remove the need for a scroll-bound event handler on video, and help with embedding video components and `amp-timeago` on Bento.

**Comment by dvoytenko** [comment]
> @dvoytenko RE: siloing to inabox, should we try to get rid of `viewportCallback` in general in favor of `IntersectionObserver` everywhere?
> 
> This would also remove the need for a scroll-bound event handler on video, and help with embedding video components and `amp-timeago` on Bento.

If we talk about the future: absolutely, I'd like to get rid of the `viewportCallback` and use `InOb` instead. However, a few things will have to change for this and that's why I recommend solving this for inabox first. Things that need to change before we can do this in the main AMP codebase are:

1. We will need to switch from a generic `viewportCallback` to a more feature-specific `InOb` and it always has to be opt-in. In other words - `viewportCallback` should no longer be called on every single element. Instead loader system could use its own `InOb` and other features could use their own.
2. We need to solve for InOb polyfill. Or, alternatively, we could define a behavior when it's absent per feature. For some features it could be the feature is disabled (e.g. loaders are disabled entirely on IE) vs some other features we assume an element is always in viewport (yeah, bad for performance, but...).

**Comment by alanorozco** [comment]
> We will need to switch from a generic viewportCallback to a more feature-specific InOb and it always has to be opt-in. In other words - viewportCallback should no longer be called on every single element. Instead loader system could use its own InOb and other features could use their own.

It sounds like we can make the opt-in API here similar to the `InOb`-specific API you mention, just removing knowledge of `viewportCallbacks` but not siloing it to inabox.

> We need to solve for InOb polyfill. Or, alternatively, we could define a behavior when it's absent per feature. For some features it could be the feature is disabled (e.g. loaders are disabled entirely on IE) vs some other features we assume an element is always in viewport (yeah, bad for performance, but...).

Fair, I believe this PR still leaves the opportunity for that to be solved in the future. FWIW, IE is left out from this workaround.

Let me know if I'm missing something about making this specific to inabox at the moment.

**Comment by alanorozco** [comment]
Discussed offline with @dvoytenko and @jridgewell. We're special-casing some component types on `InaboxResources` to keep `viewportCallback` API the same.

@lannka @dvoytenko @jridgewell

PTAL, ready for review.

File: src/inabox/inabox-resources.js [hunk 2]
```diff
@@ -56,10 +73,21 @@ export class InaboxResources {
     /** @const @private {!Deferred} */
     this.firstPassDone_ = new Deferred();
 
+    /** @private {?IntersectionObserver} */
+    this.intersectionObserver_ = null;
```
Comment (dvoytenko) [inline src/inabox/inabox-resources.js]: Be explicit. Sharing InOb instances for any generic reason is hard. Let's call it `inViewportObserver`.

File: src/inabox/inabox-resources.js [hunk 3]
```diff
@@ -183,6 +212,41 @@ export class InaboxResources {
     this.passObservable_.fire();
     this.firstPassDone_.resolve();
   }
+
+  /**
+   * @return {?IntersectionObserver}
+   * @private
+   */
+  initIntersectionObserver_() {
```
Comment (dvoytenko) [inline src/inabox/inabox-resources.js]: ditto: `iniInViewportObserver`

File: src/inabox/inabox-resources.js [hunk 4]
```diff
@@ -183,6 +212,41 @@ export class InaboxResources {
...
+    if (this.intersectionObserver_) {
+      return this.intersectionObserver_;
+    }
+    const {IntersectionObserver} = this.win;
+    if (!IntersectionObserver) {
+      return null;
+    }
+    return (this.intersectionObserver_ = new IntersectionObserver(
+      entries => entries.forEach(triggerViewportCallbackFromIntersection),
+      {threshold: 0.5}
```
Comment (dvoytenko) [inline src/inabox/inabox-resources.js]: 0.5 seems very strange. We usually do `threshold: 0` for this. No?

File: src/inabox/inabox-resources.js [hunk 5]
```diff
@@ -183,6 +212,41 @@ export class InaboxResources {
...
+    ));
+  }
+
+  /**
+   * @param {!Resource} resource
+   * @private
+   */
+  maybeObserveIntersections_(resource) {
```
Comment (dvoytenko) [inline src/inabox/inabox-resources.js]: `maybeObserveInViewport_`

File: src/inabox/inabox-resources.js [hunk 6]
```diff
@@ -183,6 +212,41 @@ export class InaboxResources {
...
+    const {element, debugid} = resource;
+    if (!triggerViewportCallbackOn.includes(element.tagName)) {
+      return;
+    }
+    const observer = this.initIntersectionObserver_();
+    if (!observer) {
+      return;
+    }
+    observer.observe(element);
```
Comment (dvoytenko) [inline src/inabox/inabox-resources.js]: There should be `unobserve` somewhere.

File: src/inabox/inabox-resources.js (hunk 4 above)
Comment (alanorozco) [reply src/inabox/inabox-resources.js]: Used for symmetry with how video autoplay works. No strong opinion either way, removed to use default `0`.

File: src/inabox/inabox-resources.js (hunk 6 above)
Comment (alanorozco) [reply src/inabox/inabox-resources.js]: Doing on `remove()`, let me know if I'm missing somewhere else.

File: src/inabox/inabox-resources.js (hunk 4 above)
Comment (dvoytenko) [reply src/inabox/inabox-resources.js]: To clarify, 0 is only b/c the main doc Resources does 0 as well.

**Pull request merged**